File: services/auth-service/app/api/v1/workspaces.py
"""Workspace routes."""

from typing import List
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.workspace import (
    WorkspaceCreate,
    WorkspaceMemberCreate,
    WorkspaceMemberResponse,
    WorkspaceResponse,
    WorkspaceUpdate,
)
from app.services.workspace_service import WorkspaceService
from app.services.event_publisher import event_publisher

router = APIRouter(prefix="/workspaces", tags=["Workspaces"])


@router.get("", response_model=List[WorkspaceResponse])
async def list_workspaces(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List all workspaces the current user has access to."""
    workspace_service = WorkspaceService(db)
    workspaces = await workspace_service.get_user_workspaces(
        current_user.id, skip=skip, limit=limit
    )
    return [WorkspaceResponse.model_validate(w) for w in workspaces]


@router.post("", response_model=WorkspaceResponse, status_code=status.HTTP_201_CREATED)
async def create_workspace(
    workspace_create: WorkspaceCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Create a new workspace."""
    workspace_service = WorkspaceService(db)

    # Check if slug already exists
    existing = await workspace_service.get_by_slug(workspace_create.slug)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Workspace with this slug already exists",
        )

    workspace = await workspace_service.create(workspace_create, current_user.id)

    # Publish event
    try:
        await event_publisher.publish(
            "workspace.created",
            {"workspace_id": workspace.id, "name": workspace.name, "owner_id": current_user.id},
            workspace_id=workspace.id,
        )
    except Exception:
        pass

    return WorkspaceResponse.model_validate(workspace)


@router.get("/{workspace_id}", response_model=WorkspaceResponse)
async def get_workspace(
    workspace_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get a specific workspace."""
    workspace_service = WorkspaceService(db)
    workspace = await workspace_service.get_by_id(workspace_id)

    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )

    # Check access
    if not await workspace_service.is_member(workspace_id, current_user.id):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have access to this workspace",
        )

    return WorkspaceResponse.model_validate(workspace)


@router.put("/{workspace_id}", response_model=WorkspaceResponse)
async def update_workspace(
    workspace_id: UUID,
    workspace_update: WorkspaceUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Update a workspace."""
    workspace_service = WorkspaceService(db)

    try:
        workspace = await workspace_service.update(
            workspace_id, workspace_update, current_user.id
        )
    except PermissionError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e),
        )

    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )

    return WorkspaceResponse.model_validate(workspace)


@router.delete("/{workspace_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_workspace(
    workspace_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Delete a workspace."""
    workspace_service = WorkspaceService(db)

    try:
        deleted = await workspace_service.delete(workspace_id, current_user.id)
    except PermissionError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e),
        )

    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )


@router.get("/{workspace_id}/members", response_model=List[WorkspaceMemberResponse])
async def list_workspace_members(
    workspace_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List all members of a workspace."""
    workspace_service = WorkspaceService(db)

    # Check access
    if not await workspace_service.is_member(workspace_id, current_user.id):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have access to this workspace",
        )

    members = await workspace_service.get_members(workspace_id)
    return [WorkspaceMemberResponse.model_validate(m) for m in members]


@router.post(
    "/{workspace_id}/members",
    response_model=WorkspaceMemberResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_workspace_member(
    workspace_id: UUID,
    member_create: WorkspaceMemberCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Add a member to a workspace."""
    workspace_service = WorkspaceService(db)

    try:
        member = await workspace_service.add_member(
            workspace_id, member_create, current_user.id
        )
    except PermissionError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e),
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )

    return WorkspaceMemberResponse.model_validate(member)


@router.delete(
    "/{workspace_id}/members/{user_id}", status_code=status.HTTP_204_NO_CONTENT
)
async def remove_workspace_member(
    workspace_id: UUID,
    user_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Remove a member from a workspace."""
    workspace_service = WorkspaceService(db)

    try:
        removed = await workspace_service.remove_member(
            workspace_id, user_id, current_user.id
        )
    except PermissionError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e),
        )

    if not removed:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Member not found",
        )

File: frontend/dashboard/src/api/workspaces.ts
/**
 * Workspaces API functions
 */
import apiClient from './client';

export interface Workspace {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  owner_id: string;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceMember {
  user_id: string;
  workspace_id: string;
  role: string;
  joined_at: string;
  email?: string;
  full_name?: string;
}

export interface Site {
  id: string;
  name: string;
  domain: string;
  workspace_id: string;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export async function getWorkspaces(): Promise<Workspace[]> {
  const response = await apiClient.get<Workspace[]>('/api/v1/workspaces');
  return response.data;
}

export async function getWorkspace(workspace_id: string): Promise<Workspace> {
  const response = await apiClient.get<Workspace>(
    `/api/v1/workspaces/${workspace_id}`
  );
  return response.data;
}

export async function createWorkspace(data: {
  name: string;
  slug: string;
  description?: string;
}): Promise<Workspace> {
  const response = await apiClient.post<Workspace>('/api/v1/workspaces', data);
  return response.data;
}

export async function updateWorkspace(
  workspace_id: string,
  data: { name?: string; description?: string }
): Promise<Workspace> {
  const response = await apiClient.put<Workspace>(
    `/api/v1/workspaces/${workspace_id}`,
    data
  );
  return response.data;
}

export async function deleteWorkspace(workspace_id: string): Promise<void> {
  await apiClient.delete(`/api/v1/workspaces/${workspace_id}`);
}

export async function getWorkspaceMembers(
  workspace_id: string
): Promise<WorkspaceMember[]> {
  const response = await apiClient.get<WorkspaceMember[]>(
    `/api/v1/workspaces/${workspace_id}/members`
  );
  return response.data;
}

export async function addWorkspaceMember(
  workspace_id: string,
  data: { email: string; role: string }
): Promise<WorkspaceMember> {
  const response = await apiClient.post<WorkspaceMember>(
    `/api/v1/workspaces/${workspace_id}/members`,
    data
  );
  return response.data;
}

export async function removeWorkspaceMember(
  workspace_id: string,
  user_id: string
): Promise<void> {
  await apiClient.delete(
    `/api/v1/workspaces/${workspace_id}/members/${user_id}`
  );
}

// Sites
export async function getSites(workspace_id: string): Promise<Site[]> {
  const response = await apiClient.get<Site[]>(
    `/api/v1/workspaces/${workspace_id}/sites`
  );
  return response.data;
}

export async function getSite(site_id: string): Promise<Site> {
  const response = await apiClient.get<Site>(`/api/v1/sites/${site_id}`);
  return response.data;
}

export async function createSite(
  workspace_id: string,
  data: { name: string; domain: string; description?: string }
): Promise<Site> {
  const response = await apiClient.post<Site>(
    `/api/v1/workspaces/${workspace_id}/sites`,
    data
  );
  return response.data;
}

export async function updateSite(
  site_id: string,
  data: { name?: string; domain?: string; description?: string }
): Promise<Site> {
  const response = await apiClient.put<Site>(`/api/v1/sites/${site_id}`, data);
  return response.data;
}

export async function deleteSite(site_id: string): Promise<void> {
  await apiClient.delete(`/api/v1/sites/${site_id}`);
}
